Optimization of `Detector.collisions` method

bchevalier opened the pull request:
### Changes
- The whole logic to keep track of starting, ending and active collisions has been simplified. It used to be spread across 2 methods: `Detector.collisions` and `Pairs.update`. It is now segregated to the `Detector.collisions` method alone (`Pairs.update` was removed).
- Collision pairs are no stored on bodies (the body with the lowest id stores the collision), it allowed to get rid of the slow map lookups and pair id generation - this was achieved efficiently by maintaining a list of bodies sorted by id on each composite.

Includes #522, #526, #527, #528

### Results
The heavier the simulation, the higher the gains. Reduced resource usage can be as high as 54.4% (Stress 2). All the measured tests seem to benefit from the optimization.

#### Stress - Performance gain: 46.6%
Before
<img width="416" alt="screen shot 2017-12-26 at 10 48 28 am" src="https://user-images.githubusercontent.com/1897225/34345707-52900e0c-ea34-11e7-9904-95a7783229df.png">

After
<img width="420" alt="screen shot 2017-12-26 at 10 47 47 am" src="https://user-images.githubusercontent.com/1897225/34345708-54f3ef74-ea34-11e7-8779-3a94eb808020.png">

#### Stress 2 - Performance gain: 54.4%
Before
<img width="421" alt="screen shot 2017-12-26 at 10 52 58 am" src="https://user-images.githubusercontent.com/1897225/34345795-1f623536-ea35-11e7-9665-f6a1866e07af.png">

After
<img width="416" alt="screen shot 2017-12-26 at 10 52 48 am" src="https://user-images.githubusercontent.com/1897225/34345794-1a78c21a-ea35-11e7-8c24-baf80a7ff143.png">

#### Bridge - Performance gain: 21.4%
Before
<img width="437" alt="screen shot 2017-12-26 at 11 01 53 am" src="https://user-images.githubusercontent.com/1897225/34345866-aec9ac54-ea35-11e7-84b7-0f6b656af6ee.png">

After
<img width="433" alt="screen shot 2017-12-26 at 11 01 46 am" src="https://user-images.githubusercontent.com/1897225/34345871-b36ef5ac-ea35-11e7-886a-6083e4421e9c.png">

#### Events - Performance gain: 18.9%
Before
<img width="430" alt="screen shot 2017-12-26 at 11 41 31 am" src="https://user-images.githubusercontent.com/1897225/34345907-fd527b80-ea35-11e7-82f4-69f700f04fd8.png">

After
<img width="432" alt="screen shot 2017-12-26 at 11 40 22 am" src="https://user-images.githubusercontent.com/1897225/34345914-05f40998-ea36-11e7-8f61-e392b99f1d39.png">


***Notes:***
- In general, the higher the number of collisions, the higher the performance gain.
- We can notice that methods such as `solveVelocity` benefit from the optimization as well as there is no need to loop over inactive pairs anymore.
- `Detector.collisions` method is running faster in all cases but the `SAT.collides` method does not benefit from gains when the number of collisions is small.
- `Pairs.update` and `Pair.id` method have disappeared
- `Grid.update` method benefits from a good performance gain in general.

### Overall results
I wanted to resume the overall performance gain achieved by the different optimizations in #522, #526, #527, #528, including this PR:

**Stress** - Performance gain: 81% (5.3 x faster)
**Stress 2** - Performance gain: 84% (6.2 x faster)
**Bridge** - Performance gain: 52% (2.1 x faster)
**Events** - Performance gain: 42% (1.7 x faster)

ronkorving commented:
Hi there, just checking in.
Are these optimization PRs still going places?

bchevalier commented:
performance gain on `Stress 2` is now about 9x!

liabru commented:
Impressive work! This is going to take some time to review, but I've tried it out and the gains appear to be huge. Looks like we'll be needing a harder stress test too :)

But there are a few issues I can see so far:

- the ragdoll demo is missing the ragdoll
- the debugger shows an extra thread in the ragdoll demo (might be related to the above bug)
- the world clear button on the GUI doesn't work now
- the boxes in the catapult demo do not fly very high anymore (not sure if that indicates anything)

bchevalier commented:
I had fixed some of the issues in later PR. I imported the specific commits. I am not sure I fixed the extra thread issue. I am not sure how I can test it?

About the catapult, the behavior before and after seem the same.
Before:
<img width="562" alt="screen shot 2018-03-24 at 5 25 52 pm" src="https://user-images.githubusercontent.com/1897225/37861952-853ecbb0-2f88-11e8-8bfa-45928ece6f21.png">


After:
<img width="552" alt="screen shot 2018-03-24 at 5 25 38 pm" src="https://user-images.githubusercontent.com/1897225/37861951-811f1d46-2f88-11e8-98e7-4aab5059000b.png">


Screenshots of maximum reached heights.

Yahasana commented:
Look promise and any things new?

NikhilVerma commented:
I am looking forward to this merge as well

morgan3d commented:
Any progress on merging these performance pull requests? All combined, they are a major improvement!
